File: src/audioWfc/wfc/constraints/DescendingMelodySoftConstraint.java
package audioWfc.wfc.constraints;

import audioWfc.wfc.HigherValues;
import audioWfc.musicTheory.OctavedNote;
import audioWfc.wfc.Tile;
import audioWfc.wfc.constraints.concepts.SoftConstraint;

public class DescendingMelodySoftConstraint extends SoftConstraint<OctavedNote> {
    public DescendingMelodySoftConstraint(double factor){
        super(factor);
    }

    @Override
    public double weight(Tile<OctavedNote> item, HigherValues higherValues) {
        OctavedNote note = item.getValue();

        Tile<OctavedNote> prev = item.getPrev();
        Tile<OctavedNote> next = item.getNext();

        double out = 0d;

        if(prev.isCollapsed() && checkDescending(prev.getValue(), note)){
            out += factor;
        }

        if(next.isCollapsed() && checkDescending(note, next.getValue())){
            out += factor;
        }

        return out;
    }

    private boolean checkDescending(OctavedNote first, OctavedNote second){
        return first.MIDIValue() > second.MIDIValue();
    }

    @Override
    public String name() {
        return "Descending melody";
    }

    @Override
    public String configText() {
        return "Bonus: " + factor;
    }
}

File: vite-project/src/wfc/constraints/DescendingMelodySoftConstraint.ts
import { OctavedNote } from "../../music_theory/Note"
import { Tile } from "../Tile"
import { SoftConstraint } from "./concepts/Constraint"
import { noteConstraintTypeToName } from "./constraintUtils"

export class DescendingMelodySoftConstraint extends SoftConstraint<OctavedNote>{
	name = noteConstraintTypeToName.get("DescendingMelodySoftConstraint") as string
	configText = () => `Bonus: ${this.bonus}`

	constructor(bonus: number) {
		super(bonus)
	}

	weight(tile: Tile<OctavedNote>): number {
		const note = tile.getValue()
		const prev = tile.getPrev()
		const next = tile.getNext()

		let out = 0

		if(prev.isCollapsed() && this.isDescending(prev.getValue(), note)) out += this.bonus
		if(next.isCollapsed() && this.isDescending(note, next.getValue())) out += this.bonus

		return out
	}

	private isDescending(first: OctavedNote, second: OctavedNote): boolean {
		return first.toMIDIValue() > second.toMIDIValue()
	}
}